spark/v3.2/spark/src/main/java/org/apache/iceberg/spark/procedures/RewriteDataFilesProcedure.java

```diff
@@ -0,0 +1,192 @@
+      new StructField("added_data_files_count", DataTypes.IntegerType, false, Metadata.empty())
+  });
+
+  public static ProcedureBuilder builder() {
+    return new Builder<RewriteDataFilesProcedure>() {
+      @Override
+      protected RewriteDataFilesProcedure doBuild() {
+        return new RewriteDataFilesProcedure(tableCatalog());
+      }
+    };
+  }
+
+  private RewriteDataFilesProcedure(TableCatalog tableCatalog) {
+    super(tableCatalog);
+  }
+
+  @Override
+  public ProcedureParameter[] parameters() {
+    return PARAMETERS;
+  }
+
+  @Override
+  public StructType outputType() {
+    return OUTPUT_TYPE;
+  }
+
+  @Override
+  public InternalRow[] call(InternalRow args) {
+    Identifier tableIdent = toIdentifier(args.getString(0), PARAMETERS[0].name());
+
+    return modifyIcebergTable(tableIdent, table -> {
+      RewriteDataFiles action = actions().rewriteDataFiles(table);
+
+      String strategy = args.isNullAt(1) ? null : args.getString(1);
+      SortOrder sortOrder = collectSortOrders(table, args.isNullAt(2) ? null : args.getString(2));
+      action = checkAndApplyStrategy(action, strategy, sortOrder);
+
+      if (!args.isNullAt(3)) {
+        action = checkAndApplyOptions(args, action);
+      }
+
+      String where = args.isNullAt(4) ? null : args.getString(4);
+      action = checkAndApplyFilter(action, where);
+
+      RewriteDataFiles.Result result = action.execute();
+
+      return toOutputRows(result);
+    });
+  }
+
+  private RewriteDataFiles checkAndApplyFilter(RewriteDataFiles action, String where) {
+    if (where != null) {
+      ParserInterface sqlParser = spark().sessionState().sqlParser();
+      try {
+        Expression expression = sqlParser.parseExpression(where);
+        return action.filter(SparkExpressionConverter.convertToIcebergExpression(expression));
+      } catch (ParseException e) {
+        throw new IllegalArgumentException("Cannot parse predicates in where option: " + where);
+      }
+    }
+    return action;
+  }
+
+  private RewriteDataFiles checkAndApplyOptions(InternalRow args, RewriteDataFiles action) {
+    Map<String, String> options = Maps.newHashMap();
+    args.getMap(3).foreach(DataTypes.StringType, DataTypes.StringType,
+            (k, v) -> {
+              options.put(k.toString(), v.toString());
+              return BoxedUnit.UNIT;
+            });
+    return action.options(options);
+  }
+
+  private RewriteDataFiles checkAndApplyStrategy(RewriteDataFiles action, String strategy, SortOrder sortOrder) {
+    if (strategy != null) {
+      if (strategy.equalsIgnoreCase("binpack")) {
+        RewriteDataFiles rewriteDataFiles = action.binPack();
+        if (sortOrder != null) {
+          // calling below method to throw the error as user has set both binpack strategy and sortorder
+          return rewriteDataFiles.sort(sortOrder);
+        }
+        return rewriteDataFiles;
+      } else if (strategy.equalsIgnoreCase("sort")) {
```

Comment:
could just be this || sortOrder.nonEmpty right?

Reply:
sortOrder is not a string here, so we don't have sortOrder.nonEmpty right?

Reply:
What i'm suggesting here is that we want to enter this branch in one of two conditions, currently they are in separate branches.
```
If binpack strategy {
  // binpack
} else if sort strategy {
  // Sort
} else { 
  if Sort Order Set {
    // sort
  }
}
```  

When we can have it be
```
If binpack strategy {
  // binpack
} else if sort strategy || sort order set{
  // Sort
}
```

Reply:
The check to make sure you don't have binpack and order set at the same time should be part of the validation for args I think.

Reply:
https://github.com/apache/iceberg/pull/3375#discussion_r743791976

I initially added in the args validation, but as per the above comment I removed it.

Reply:
I think it looks messy because of below branch

```
if (strategy.equalsIgnoreCase("binpack")) {
      RewriteDataFiles rewriteDataFiles = action.binPack();
      if (sortOrder != null) {
        // calling below method to throw the error as user has set both binpack strategy and sortorder
        return rewriteDataFiles.sort(sortOrder);
      }
      return rewriteDataFiles;
    }
```

If strategy=binpack and sortorder is set, I am using the  BaseAction to throw error, instead of new code throwing error. It was suggested  here [#3375 (comment)](https://github.com/apache/iceberg/pull/3375#discussion_r743791976)

Reply:
The above link doesn't work for me, but the issue here is this is one of the cases where the underlying action cannot validate the options. Since the apis are either
```
.binPack
.sort 
.sort(sort order)
```
You cannot normally set binpack with a sort order so I think we probably should not allow it here either.

Reply:
oh got it. 
action.sort() and action.sort(null) is one and the same. So, no need to explicitly check for sortorder is null or not. I can always pass it.

Reply:
I have simplified now.